File: android/src/main/java/app/globy/zettle/ZettlePlugin.kt
package app.globy.zettle

import android.app.Activity
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.zettle.sdk.ZettleSDK
import com.zettle.sdk.feature.cardreader.ui.CardReaderAction
import com.zettle.sdk.feature.cardreader.ui.payment.CardPaymentResult
import com.zettle.sdk.ui.ZettleResult
import com.zettle.sdk.ui.zettleResult
import kotlin.math.roundToLong


@CapacitorPlugin(name = "Zettle")
class ZettlePlugin : Plugin() {

    private var zettle: Zettle? = null;

    override fun load() {
        super.load()

        activity?.let { zettle = Zettle(it) }
    }

    @PluginMethod
    fun initialize(call: PluginCall?) {
        val devmode = call?.getBoolean("developermode", false) == true
        val clientId = config.getString("clientID")
        if (clientId == null) {
            call?.reject("Must set a clientID in the configuration.")
            return
        }
        val scheme = config.getString("scheme")
        if (scheme == null) {
            call?.reject("Must set a scheme in the configuration.")
            return
        }
        val host = config.getString("host")
        if (host == null) {
            call?.reject("Must set a host in the configuration.")
            return
        }

        zettle?.initialize(clientId, scheme, host, devmode)
    }

    @PluginMethod
    fun chargeAmount(call: PluginCall?) {
        val amount = call?.getDouble("amount")
        if (amount == null) {
            call?.reject("Must provide amount")
            return
        }

        val reference = call.getString("reference")

        val intent = zettle?.getPaymentIntent((amount * 100).roundToLong(), reference)
        startActivityForResult(call, intent, "finishPayment")
    }

    @ActivityCallback
    private fun finishPayment(call: PluginCall?, activityResult: ActivityResult) {
        if (call == null) {
            return
        }

        if (activityResult.resultCode != Activity.RESULT_OK) {
            return
        }

        when (val result = activityResult.data?.zettleResult()) {
            is ZettleResult.Completed<*> -> {
                val payment: CardPaymentResult.Completed = CardReaderAction.fromPaymentResult(result)
                val ret = JSObject()
                ret.put("success", true)
                ret.put("amount", payment.payload.amount)
                ret.put("gratuityAmount", payment.payload.gratuityAmount)
                ret.put("referenceNumber", payment.payload.referenceNumber)
                ret.put("entryMode", payment.payload.cardPaymentEntryMode)
                ret.put("obfuscatedPan", payment.payload.maskedPan)
                ret.put("transactionId", payment.payload.transactionId)
                ret.put("cardBrand", payment.payload.cardType)
                ret.put("authorizationCode", payment.payload.authorizationCode)
                ret.put("AID", payment.payload.applicationIdentifier)
                ret.put("TSI", payment.payload.tsi)
                ret.put("TVR", payment.payload.tvr)
                ret.put("applicationName", payment.payload.applicationName)
                ret.put("numberOfInstallments", payment.payload.nrOfInstallments)
                ret.put("installmentAmount", payment.payload.installmentAmount)
                call.resolve(ret)
            }
            else -> {
                // Errors are not documented well.
                // Our implementation choose to just collapse it into a { success: false } object for brevity
                // See https://github.com/iZettle/sdk-ios/issues/470
                val response = JSObject();
                response.put("success", false)
                call.resolve(response)
            }
        }
    }

    @PluginMethod
    fun logout(call: PluginCall?) {
        zettle?.logout()
    }

    @PluginMethod
    fun openSettings(call: PluginCall?) {
        zettle?.showCardReaderSettings(this.bridge.activity)
    }
}

File: ios/Sources/ZettlePlugin/ZettlePlugin.swift
import Foundation
import Capacitor

import iZettleSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ZettlePlugin)
public class ZettlePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ZettlePlugin"
    public let jsName = "Zettle"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnNone),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnNone),
        CAPPluginMethod(name: "openSettings", returnType: CAPPluginReturnNone),
        CAPPluginMethod(name: "chargeAmount", returnType: CAPPluginReturnPromise)
    ]

    private let impl = Zettle()

    private var callbackId: String?

    @objc func initialize(_ call: CAPPluginCall) {
        let developerMode = call.getBool("developerMode")

        guard let clientID = getConfig().getString("clientID") else { call.reject("Must set a clientID in the configuration."); return }
        guard let host = getConfig().getString("host")
        else { call.reject("Must set a scheme in the configuration."); return }
        guard let scheme = getConfig().getString("scheme")
        else { call.reject("Must set a host in the configuration."); return}

        guard let callbackURL = URL(string: scheme + "://" + host) else { return }

        impl.initialize(clientID: clientID, callbackURL: callbackURL, enableDeveloperMode: developerMode ?? false)
    }

    @objc func logout(_ call: CAPPluginCall) {
        impl.logout()
    }

    @objc func openSettings(_ call: CAPPluginCall) {
        guard let bridge = self.bridge else { return }
        guard let viewController = bridge.viewController else { return }

        impl.presentSettings(viewController: viewController)
    }

    @objc func chargeAmount(_ call: CAPPluginCall) {
        guard let bridge = self.bridge else { return }
        guard callbackId == nil else { call.reject("A payment was already started"); return }
        bridge.saveCall(call)
        callbackId = call.callbackId

        guard let amount = call.getDouble("amount") else { call.reject("Must provide an amount"); return }

        let reference = call.getString("reference")

        guard let bridge = self.bridge else { return }
        guard let viewController = bridge.viewController else { return }

        impl.chargeAmount(
            viewController: viewController,
            amount: NSDecimalNumber(decimal: Decimal(amount)),
            reference: reference,
            callback: paymentCompletion(paymentInfo:error:)
        )

    }

    private func paymentCompletion(paymentInfo: iZettleSDKPaymentInfo?, error: (any Error)?) {
        guard let call = self.bridge?.savedCall(withID: callbackId ?? "") else { return }

        if let nsError = error as? NSError {
            // Errors are not documented well.
            // Our implementation choose to just collapse it into a { success: false } object for brevity
            // See https://github.com/iZettle/sdk-ios/issues/470
            call.resolve(["success": false])
        } else {
            // Payment was unsuccesfull
            guard let payment = paymentInfo else {
                call.resolve(["success": false])
                return }

            call.resolve([
                "success": true,
                "amount": payment.amount,
                "gratuityAmount": payment.gratuityAmount as Any,
                "referenceNumber": payment.referenceNumber,
                "entryMode": payment.entryMode,
                "obfuscatedPan": payment.obfuscatedPan,
                "panHash": payment.panHash,
                "transactionId": payment.transactionId,
                "cardBrand": payment.cardBrand,
                "authorizationCode": payment.authorizationCode,
                "AID": payment.aid as Any,
                "TSI": payment.tsi as Any,
                "TVR": payment.tvr as Any,
                "applicationName": payment.applicationName as Any,
                "numberOfInstallments": payment.numberOfInstallments as Any,
                "installmentAmount": payment.installmentAmount as Any
            ])
        }

        callbackId = nil
    }
}
